src/NServiceBus.Core/Routing/MessageDrivenSubscriptions/Publishers.cs

```diff
@@ -12,8 +12,28 @@ public class Publishers
     {
         internal IEnumerable<PublisherAddress> GetPublisherFor(Type eventType)
         {
-            var distinctPublishers = rules.Select(r => r.Apply(eventType)).Where(e => e != null).Distinct().ToList();
-            return distinctPublishers;
+            List<PublisherAddress> staticPublishersForType;
+            if (!staticPublishers.TryGetValue(eventType, out staticPublishersForType))
+            {
+                staticPublishersForType = emptyList;
+            }
+
+            if (dynamicRules.Count > 0)
+            {
+                var dynamicAddresses = new List<PublisherAddress>();
+                foreach (var rule in dynamicRules)
+                {
+                    var address = rule(eventType);
+                    if (address != null)
+                    {
+                        dynamicAddresses.Add(address);
+                    }
+                }
+
+                return staticPublishersForType.Count > 0 ? staticPublishersForType.Concat(dynamicAddresses) : dynamicAddresses;
+            }
+
+            return staticPublishersForType;
```

Comment:
Of course, given that `PublisherAddress` doesn't have any equality overloads, we were only getting reference equality checking with the previous `Distinct` call anyway.

So, if we implemented the appropriate bits in `PublisherAddress`, we could then store them here in a HashSet, and get value deduping.

Or, we can just keep it as is, and leave deduping up to the caller once they've resolved the addresses to strings!